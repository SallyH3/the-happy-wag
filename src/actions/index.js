export const storeAnimals = (animals) => ({
  type: 'STORE_ANIMALS',
  animals
})
