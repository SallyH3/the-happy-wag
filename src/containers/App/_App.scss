.main-loader {
  font-family: 'Mandali', sans-serif;
  text-align: center;
  font-size: 30px;
}