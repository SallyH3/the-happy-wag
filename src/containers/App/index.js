import React, { Component }from 'react';
import './_App.scss';
import { apiKey } from '../../utils/apikey.js';
import Header from '../../components/Header/index.js';
import CardWrapper from '../CardWrapper/index.js';
import CardDetails from '../../components/CardDetails/index.js';
import { connect } from 'react-redux';
import { Route } from 'react-router-dom';
import { storeAnimals } from '../../actions';
import './_App.scss';

class App extends Component {
  constructor() {
    super();
    this.state = {
      token: '',
      animals: [],
      isLoading: true
    }
  }

  fetchToken = () => {
    fetch('http://localhost:3001')
    .then(response => response.json())
    .then(credentials => this.setState({ token: credentials.token }, this.fetchAnimals))
  }

  fetchAnimals = () => {
    const options = {
      headers: {
        Authorization: 'Bearer ' + this.state.token
      }
    }
    fetch('https://api.petfinder.com/v2/animals', options)
    .then(response => response.json())
    .then(results => this.setState({ animals: results.animals, isLoading: false }))
  }

  componentDidMount = () => {
    this.setState({ isLoading: true })
    this.fetchToken()
  }



  render() {
    let result;
    if(this.state.isLoading === true) {
      result = <p className='main-loader'>Loading - please wait...</p>
    } 
    return (
      <section className='App'>
        <Route exact path='/' component= { Header } />
          {result}
        <CardWrapper animals={this.state.animals}/>
        <Route to='/CardDetails' component={ CardDetails } />
        <Route path='/CardDetails/:id' render={({ match }) => {
          const { id } = match.params;
          const selectedCard = this.state.animals.find(animal => {
            return animal.animal_id === parseInt(id)
          })
          if(selectedCard) {
            return <CardDetails {...selectedCard} />
          }
        }}
        />  
      </section>
    )
  }
}

export const mapStateToProps = (state) => ({
  animals: state.animals
})

export const mapDispatchToProps = (dispatch) => ({
  addAnimals: animals => dispatch(storeAnimals(animals))
})

export default connect(mapStateToProps, mapDispatchToProps)(App)
