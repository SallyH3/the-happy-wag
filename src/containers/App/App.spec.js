import { App, mapStateToProps, mapDispatchToProps } from '../App';
import { shallow } from 'enzyme';
import React from 'react';
import { setAnimals, handleLoading } from '../../actions';
import { createSerializer } from 'enzyme-to-json';

expect.addSnapshotSerializer(createSerializer({mode: 'deep'}));

describe('App', () => {
	let wrapper;

	beforeEach(() => {
		wrapper = shallow(
				<App />
		);
  });
  
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  })

  describe('mapStateToProps', () => {

		it('should return an object with the animals array', () => {

			const mockState = {
				animals: ['animals']
			}

			const expected = {
				animals: mockState.animals 
			}

			const mappedProps = mapStateToProps(mockState)
			expect(mappedProps).toEqual(expected)
    })

    it('should return a boolean if the page is loading or data has been fetched', () => {
      const mockState = {
        boolean: true
      }

      const expected = {
        boolean: mockState.boolean
      }

      const mappedProps = mapStateToProps(mockState)

      expect(mappedProps).toEqual(expected)
    })
  });
  
  describe('mapDispatchToProps', () => {

    it('should call dispatch when using a function, setAnimals from mapDispatchToProps', () => {

      const mockDispatch = jest.fn()

      const actionToDispatch = setAnimals(['animals'])

      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.setAnimals(['animals'])
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should call dispatch when using a function, handleLoading from mapDispatchToProps', () => {

      const mockDispatch = jest.fn()

      const actionToDispatch = handleLoading(true)

      const mappedProps = mapDispatchToProps(mockDispatch)

      mappedProps.isLoading(true)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})