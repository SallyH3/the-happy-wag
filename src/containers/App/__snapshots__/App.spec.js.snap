// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<section
  className="App"
>
  <Route
    component={[Function]}
    exact={true}
    path="/"
  />
  <CardWrapper />
  <Route
    exact={true}
    path="/CardDetails/:id"
    render={[Function]}
  />
</section>
`;
